### DISCUSSION TITLE
how to get the client Ip address in app router?

### DISCUSSION THREAD

**Discussion by Arunjenson-ss** [vercel/next.js#55037] [App Router] [answered]
how to get the client Ip address in app router?

**Comment by devjiwonchoi** [comment]
Take a look at https://next-js-playground-dusky.vercel.app/ and https://next-js-playground-dusky.vercel.app/api

### Option 1 - By Route Handler
```ts
// route.ts

import { NextRequest, NextResponse } from 'next/server'

export async function GET(req: NextRequest) {
  const ip = (req.headers.get('x-forwarded-for') ?? '127.0.0.1').split(',')[0]
  return NextResponse.json({ ip })
}
```

### Option 2 - By Page Component

```ts
// page.tsx

import { headers } from 'next/headers'

export default function Home() {
  const header = headers()
  const ip = (header.get('x-forwarded-for') ?? '127.0.0.1').split(',')[0]
  // ...
}
```

Basically, the method is to get `x-forwarded-for` from the headers.

### Source

https://github.com/devjiwonchoi/next.js-playground/tree/main/app-router

### Reference

- [X-Forwared-For](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)

**Reply by leerob** [reply]
Thank you for this!

**Comment by leerob** [comment, marked as answer]
Updated the docs here to make it more clear this depends on the infra provider, but you can also read some things from headers if provided: https://github.com/vercel/next.js/pull/59719

**Reply by IDJGILL** [reply]
request.geo is removed in Next 15.

**Reply by leerob** [reply]
https://nextjs.org/docs/app/building-your-application/upgrading/version-15#nextrequest-geolocation

**Reply by callumfrance** [reply]
> If you are using Vercel, you can alternatively use the geolocation and ipAddress functions from [@vercel/functions](https://vercel.com/docs/functions/vercel-functions-package) instead

So, the recommended solution only caters to a Vercel deployment now that those properties were deprecated?

**Reply by leerob** [reply]
These properties have always required your infrastructure provider to look at the framework outputs and add the information. It was never possible for us to embed geolocation APIs into Next.js itself – it is provider specific.

**Reply by IDJGILL** [reply]
@leerob Make sense.

**Comment by arisonx** [comment]
# Passing Headers Through Middleware in Next.js with Safe-Actions

This implementation of mine worked:

You should use an **actions client (safe-actions)** because it has an integrated middleware that helps pass headers from one middleware to another until it reaches the API.

## Next.js Default Middleware Example:

```ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

export function middleware(request: NextRequest) {
 // Clone the request headers so that we don't modify the original headers object
  const requestHeaders = new Headers(request.headers);

  // Check if the hosting platform provides the client's IP address and store it in a variable
  const ip = request.headers.get("x-forwarded-for") || request.headers.get("x-real-ip") || "";
  const userAgent = request.headers.get("user-agent") || "";

  // Set the IP and user agent headers into request to be used in the action
  requestHeaders.set("x-forwarded-for", ip);
  requestHeaders.set("user-agent", userAgent);

  return NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  });

  return NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  });
}
```

## 🔹 Safe-Actions Middleware:

```ts
import { createSafeActionClient } from "next-safe-action";
import { headers } from "next/headers";

// .use is the middleware function
export const actionClient = createSafeActionClient().use(async ({ next }) => {
  const ip = (await headers()).get("x-forwarded-for");
  const userAgent = (await headers()).get("user-agent");
  return next({
    ctx: {
      ip,
      userAgent,
    },
  });
});
```

## The next step is to take these headers and pass them in your fetch request inside your action (ignore the current context code as it is being used for testing):



```ts
"use server";
import { actionClient } from "@/lib/safe-action";
import { CreateUserSchema } from "@/@types/schemas/users/CreateUserSchema";
import { z } from "zod";

const UserRegisterActionOutputSchema = z.object({
  error: z.boolean(),
  status: z.number(),
  message: z.string(),
  confirmationId: z.string().optional(),
});

export const myAction = actionClient
  .schema(CreateUserSchema)
  .outputSchema(UserRegisterActionOutputSchema)
  .action(async ({ parsedInput, ctx }) => {
    const url = `api/your-route`;
    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-forwarded-for": ctx.ip ?? "",
          "user-agent": ctx.userAgent ?? "",
        },
        body: JSON.stringify(parsedInput),
      });
    console.log(response)
    }
  });

```
## Finally, we capture the headers in the API route:

```ts
import { NextRequest, NextResponse } from "next/server";

export async function POST(request: NextRequest) {
  const userIp = request.headers.get("x-forwarded-for");
  const userIpFormatted = userIp ? userIp.split(",")[0] : null;
  const userAgent = request.headers.get("user-agent");
  const others =
    request.headers.get("x-forwarded-for")?.split(",")[0] ||
    request.headers.get("x-real-ip") ||
    request.headers.get("x-forwarded-for")?.split(",")[0] ||
    request.headers.get("x-real-ip") ||
    "IP not found";
    const data = {
    userIp: userIp,
    userAgent: userAgent,
    userIpFormatted: userIpFormatted,
    others:ip
    }

   console.log(data);

  return new Response (JSON.Stringfy(data), {
    status:200
  });
}
```


## And then:

![Captura de tela 2025-03-10 164446](https://github.com/user-attachments/assets/29af8315-dd3b-4baf-97c6-ff9e5942fb3c)


## Links:
https://next-safe-action.dev/docs/define-actions/middleware
https://httptoolkit.com/blog/what-is-x-forwarded-for/
